### PR TITLE
Improve taskbar jumplist

### LINKED ISSUE #1467: Drives aren't added to the taskbar jumplist

### PR DESCRIPTION
**Pull request by Jaiganeshkumaran**
**Resolved / Related Issues**
Items resolved / related issues by this PR.
- Closes #1467

**Details of Changes**
Add details of changes here.
- The most recent item will now be shown on the top of the taskbar jumplist.
- Fixed an issue where the root of a drive won't be added on the taskbar jumplist.

**Validation**
How did you test these changes?
- [x] Built and ran the app
- [ ] Tested the changes for accessibility

### CODE REVIEW

File: Files.sln [hunk 1]
```diff
@@ -156,7 +156,6 @@ Global
 		{A2FF3F3B-8EBC-4108-B99D-1476B7876656}.Debug|arm64.ActiveCfg = Debug|arm64
 		{A2FF3F3B-8EBC-4108-B99D-1476B7876656}.Debug|arm64.Build.0 = Debug|arm64
 		{A2FF3F3B-8EBC-4108-B99D-1476B7876656}.Debug|x64.ActiveCfg = Debug|x64
-		{A2FF3F3B-8EBC-4108-B99D-1476B7876656}.Debug|x64.Build.0 = Debug|x64
```
Comment (yaichenbaum) [inline Files.sln:159]: Can you revert this change?

File: Files/App.xaml.cs [hunk 2]
```diff
@@ -246,6 +246,7 @@ protected override async void OnLaunched(LaunchActivatedEventArgs e)
                 {
                     if (!(string.IsNullOrEmpty(e.Arguments) && MainPageViewModel.AppInstances.Count > 0))
                     {
+
                         await MainPageViewModel.AddNewTabByPathAsync(typeof(PaneHolderPage), e.Arguments);
```
Comment (yaichenbaum) [inline Files/App.xaml.cs:250]:
```suggestion
                    {
                        await MainPageViewModel.AddNewTabByPathAsync(typeof(PaneHolderPage), e.Arguments);
```

File: Files.sln (hunk 1 above)
Comment (Jaiganeshkumaran) [reply Files.sln]: Done.

File: Files/ViewModels/MainPageViewModel.cs [hunk 3]
```diff
@@ -182,6 +182,9 @@ public static async Task AddNewTabByPathAsync(Type type, string path, int atInde
                 path = "Home".GetLocalized();
             }
 
+            // Support drives launched through jumplist by stripping away the question mark at the endy.
```
Comment (duke7553) [inline Files/ViewModels/MainPageViewModel.cs:185]:
```suggestion
            // Support drives launched through jump list by stripping away the question mark at the end.
```
minor nit

File: Files/ViewModels/MainPageViewModel.cs [hunk 4]
```diff
@@ -182,6 +182,9 @@ public static async Task AddNewTabByPathAsync(Type type, string path, int atInde
                 path = "Home".GetLocalized();
             }
 
+            // Support drives launched through jump list by stripping away the question mark at the end.
+            if (path.EndsWith("\\?")) path = path.Remove(path.Length - 1);
```
Comment (gave92) [inline Files/ViewModels/MainPageViewModel.cs:186]:
```cs
// Support drives launched through jump list by stripping away the question mark at the end.
if (path.EndsWith("\\?"))
{
    path = path.Remove(path.Length - 1);
}
```

File: Files/Helpers/JumpListManager.cs [hunk 5]
```diff
@@ -59,42 +60,52 @@ private void AddFolder(string path)
...
-                    var localSettings = ApplicationData.Current.LocalSettings;
-                    displayName = localSettings.Values.Get("RecycleBin_Title", "Recycle Bin");
-                }
-                else if (App.LibraryManager.TryGetLibrary(path, out LibraryLocationItem library))
-                {
-                    var libName = Path.GetFileNameWithoutExtension(library.Path);
-                    switch (libName)
+
+                if (displayName == null)
+				{
```
Comment (gave92) [inline Files/Helpers/JumpListManager.cs:72]: Align "{"

File: Files/Helpers/JumpListManager.cs [hunk 6]
```diff
@@ -59,42 +60,52 @@ private void AddFolder(string path)
         {
             if (instance != null && !JumpListItemPaths.Contains(path))
             {
-                string displayName;
-                if (path.Equals(CommonPaths.DesktopPath, StringComparison.OrdinalIgnoreCase))
+                string displayName = null;
+                if (path.EndsWith("\\"))
                 {
-                    displayName = "ms-resource:///Resources/SidebarDesktop";
+                    // Jumplist item argument can't end with a slash so append a character that can't exist in a directory name to support listing drives.
```
Comment (gave92) [inline Files/Helpers/JumpListManager.cs:66]:
```
// Jumplist item argument can't end with a slash so append a character that can't exist in a directory name to support listing drives.
var drive = App.DrivesManager.Drives.Where(drive => drive.Path == path).FirstOrDefault();
if (drive == null)
{
    return;
}
displayName = drive.Text;
path += '?';
```

File: Files/Helpers/JumpListManager.cs [hunk 7]
```diff
@@ -59,42 +60,52 @@ private void AddFolder(string path)
...
+                        }
+                    }
+                    else
+                    {
+                        displayName = Path.GetFileName(path);
                     }
-                }
-                else
-                {
-                    displayName = Path.GetFileName(path);
```
Comment (gave92) [inline Files/Helpers/JumpListManager.cs:97]: Unrelated to the PR, but here we should likely keep the newer visited folders at the top of the list or they won't be shown if there are more that ~15 items, like this:
```cs
instance.Items.Remove(instance.Items.FirstOrDefault(x => x.Arguments.Equals(path, StringComparison.OrdinalIgnoreCase)));
instance.Items.Insert(0, jumplistItem); // Keep newer items at the top
JumpListItemPaths.Remove(JumpListItemPaths.FirstOrDefault(x => x.Equals(path, StringComparison.OrdinalIgnoreCase)));
JumpListItemPaths.Add(path);
```

**Review by gave92 — COMMENTED** [review]
Works nicely 👍. See a couple of comments.

File: Files/Helpers/JumpListManager.cs [hunk 8]
```diff
@@ -1,5 +1,6 @@
 using Files.Common;
 using Files.Filesystem;
+using ICSharpCode.SharpZipLib.Zip.Compression.Streams;
```
Comment (gave92) [inline Files/Helpers/JumpListManager.cs:3]: Remove unnecessary import

File: Files/Helpers/JumpListManager.cs [hunk 9]
```diff
@@ -59,50 +60,70 @@ private void AddFolder(string path)
         {
             if (instance != null && !JumpListItemPaths.Contains(path))
```
Comment (gave92) [inline Files/Helpers/JumpListManager.cs:61]: Forgot this one: it should be only `if (instance != null)` so if you visit a path again it will be moved to the top of the list (it will still appear only once).

File: Files/Helpers/JumpListManager.cs [hunk 10]
```diff
@@ -59,50 +60,70 @@ private void AddFolder(string path)
...
             {
-                string displayName;
-                if (path.Equals(CommonPaths.DesktopPath, StringComparison.OrdinalIgnoreCase))
+                string displayName = null;
+                if (path.EndsWith("\\"))
                 {
-                    displayName = "ms-resource:///Resources/SidebarDesktop";
+                    // Jumplist item argument can't end with a slash so append a character that can't exist in a directory name to support listing drives.
+                    var drive = App.DrivesManager.Drives.Where(drive => drive.Path == path).FirstOrDefault();
+		    if (drive == null)
```
Comment (gave92) [inline Files/Helpers/JumpListManager.cs:68]: Could you check the formatting?
image

**Review by gave92 — CHANGES_REQUESTED** [review]

File: Files/Helpers/JumpListManager.cs (hunk 10 above)
Comment (Jaiganeshkumaran) [reply Files/Helpers/JumpListManager.cs:68]: This is the problem when using the GitHub web editor. I'll fix it.

**Comment by d2dyno1** [comment]
@Jaiganeshkumaran

File: Files/Helpers/JumpListManager.cs [hunk 11]
```diff
@@ -10,133 +10,153 @@
 
 namespace Files.Helpers
 {
-    public sealed class JumpListManager
-    {
-        private JumpList instance = null;
-        private List<string> JumpListItemPaths { get; set; }
+	public sealed class JumpListManager
```
Comment (yaichenbaum) [inline Files/Helpers/JumpListManager.cs:13]: Can you fix the indentation?

**Review by yaichenbaum — CHANGES_REQUESTED** [review]

**Review by yaichenbaum — DISMISSED** [review]
Nice work!

**Review by gave92 — DISMISSED** [review]

**Comment by yaichenbaum** [comment]
@Jaiganeshkumaran can you resolve the merge conflicts?

**Review by gave92 — APPROVED** [review]

**Comment by yaichenbaum** [comment]
@Jaiganeshkumaran thank you!
